### PR TITLE
ci: move most of the prepare config into scripts

### PR DESCRIPTION
**Pull request by emilyalbini**
This PR moves most of the configuration from the CI yamls into bash scripts, driven by a small Python script (which understands and emulates the two `##vso[` commands we use).

There are two reasons why we'd want to do this:

* Being able to prepare the build environment locally by just running `src/ci/prepare.py` simplifies a lot setting up a local VM similar to CI (software pre-installed in the CI images won't be prepared, but it's a start anyway).
* When we'll switch to GitHub Actions we'll need to either duplicate code in multiple workflows or write a preprocessor. Having all the prepare steps in a single one is going to simplify the implementation of both options.

Along with the move I did a few changes to the actual scripts:

* Mirrored all the remaining external URLs we download (except chocolatey) to the `rust-lang-ci-mirrors` bucket, to increase reliability and reduce the chance of supply chain attacks. I didn't audit and mirror the CI scripts outside this PR though.
* Extracted CI-specific behavior (like issuing `##vso[` commands and detecting the host platform) into `shared.sh` and included it in most of the scripts. This way a switch to another CI provider will be less painful.

It's possible (and easier) to review this commit-by-commit.
r? @alexcrichton 
cc @rust-lang/infra

### CODE REVIEW

File: src/ci/scripts/install-wix.sh [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+# We use the WIX toolset to create combined installers for Windows, and these
+# binaries are downloaded from https://github.com/wixtoolset/wix3 originally
+
+set -euo pipefail
+IFS=$'\n\t'
+
+source "$(cd "$(dirname "$0")" && pwd)/../shared.sh"
+
+if isWindows; then
+    curl -O https://rust-lang-ci-mirrors.s3-us-west-1.amazonaws.com/rustc/wix311-binaries.zip
```
Comment (fbstj) [inline src/ci/scripts/install-wix.sh]:
```suggestion
    curl -O "${MIRRORS_BASE}/wix311-binaries.zip"
```
something similar can probably be done throughout many files?

Comment (pietroalbini) [reply src/ci/scripts/install-wix.sh]: Yeah, good catch.

**Comment by alexcrichton** [comment]
@bors: r+

Thanks for this @pietroalbini!

**Comment by pietroalbini** [comment]
@bors rollup=never

At least for linux we're sure it works (the PR builders were fine), but there is a chance I forgot to copy something/typoed stuff for the other platforms, and I don't want this to affect a rollup.

**Comment by kennytm** [comment]
`aws: command not found` on Windows.

**Comment by pietroalbini** [comment]
This wasn't actually awscli missing, but the whole prepare step failed to run as it used `#!/usr/bin/env python3`, while on the Windows agents only `python.exe` is present. Added a new step to copy `python.exe` to `python3.exe` before executing the prepare script.

@bors r=alexcrichton

**Comment by pietroalbini** [comment]
Ok getting `prepare.py` working on Windows was harder than expected. There were small little issues I fixed, but installing msys2 from chocolatey just hanged after finishing. It's not worth my time continiung to investigating making `prepare.py` work, so I changed the PR to just extract everything to scripts.

I think the extraction is still worth it without consolidating steps with `prepare.py`.

r? @alexcrichton again?

**Comment by pietroalbini** [comment]
Ok this should *finally* work. p=1 it as it's blocking a lot of infra work.

@bors r=alexcrichton rollup=never p=1

**Comment by pietroalbini** [comment]
Rebased on top of master, and fixed the failure.

@bors r=alexcrichton p=10

File: src/ci/scripts/install-msys2-packages.sh [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+IFS=$'\n\t'
+
+source "$(cd "$(dirname "$0")" && pwd)/../shared.sh"
+
+if isWindows; then
+    pacman -S --noconfirm --needed base-devel ca-certificates make diffutils tar
+
+    # FIXME(#65767): workaround msys bug, step 2
+    pacman -U --noconfirm --noprogressbar mingw-w64-$arch-ca-certificates-20180409-1-any.pkg.tar.xz
+    rm mingw-w64-$arch-ca-certificates-20180409-1-any.pkg.tar.xz
```
Comment (mati865) [inline src/ci/scripts/install-msys2-packages.sh]: Are you sure `arch` is set in this script?

Comment (pietroalbini) [reply src/ci/scripts/install-msys2-packages.sh]: Woops.

File: src/ci/scripts/install-wix.sh [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+
+source "$(cd "$(dirname "$0")" && pwd)/../shared.sh"
+
+if isWindows; then
+    curl -O "${MIRRORS_BASE}/wix311-binaries.zip"
+    mkdir -p wix/bin
+    cd wix/bin
+    7z x ../../wix311-binaries.zip
+
+    ciCommandSetEnv WIX "$(pwd)/wix"
```
Comment (mati865) [inline src/ci/scripts/install-wix.sh]: CI error was caused by this.
Previously `WIX` was set before `cd` now it's set after it.

Comment (pietroalbini) [reply src/ci/scripts/install-wix.sh]: Yep, already pushed a commit with a fix and started a try build on the failed builder.

**Comment by pietroalbini** [comment]
*Finally* :tada:

**Pull request merged**